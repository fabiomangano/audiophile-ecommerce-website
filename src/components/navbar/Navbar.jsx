import React, { useState, useCallback } from 'react';
import { Link } from 'react-router-dom';
import { useMediaQuery } from 'react-responsive';
import { ReactComponent as Logo } from '../../assets/shared/desktop/logo.svg';
import { MediumOrSmallScreen, is1440orLarger } from '../../mediaQueries';
import IconButton from '../icon-button/IconButton';
import NavLinks from '../nav-links/NavLinks';
import Icon from '../icon/Icon';
import icons from '../icon/icons';
import style from './Navbar.module.scss';

const Navbar = () => {
  const [openMenu, setOpenMenu] = useState(false);
  const [openCart, setOpenCart] = useState(false);

  const handleOpenMenu = useCallback(() => {
    setOpenMenu(!openMenu);
  }, [openMenu]);

  const handleOpenCart = useCallback(() => {
    setOpenCart(!openCart);
  }, [openCart]);

  const isMediumOrSmallScreen = useMediaQuery(MediumOrSmallScreen);
  const is1440 = useMediaQuery(is1440orLarger);

  return (
    <div className={style.container}>
      <nav>
        {isMediumOrSmallScreen && (
          <IconButton data-testid="menu-btn" onClick={handleOpenMenu}>
            {openMenu ? (
              <Icon path={icons.X} color="#fff" size={30} />
            ) : (
              <Icon path={icons.MENU} color="#fff" size={30} />
            )}
          </IconButton>
        )}

        <Link to="/">
          <Logo />
        </Link>

        {isMediumOrSmallScreen ? (
          <div
            data-testid="navlinks-container"
            className={openMenu ? style.display : style.hide}
          >
            <NavLinks direction="vertical" />
          </div>
        ) : (
          <NavLinks direction="horizontal" />
        )}

        <span className={is1440 ? style.center : undefined}>
          <IconButton data-testid="cart-btn" onClick={handleOpenCart}>
            <Icon path={icons.CART} color="#fff" />
          </IconButton>
        </span>
      </nav>
    </div>
  );
};

export default Navbar;
