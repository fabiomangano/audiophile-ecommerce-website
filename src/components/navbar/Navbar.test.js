import React from 'react';
import ShallowRender from 'react-test-renderer/shallow';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { useMediaQuery } from 'react-responsive';

import Navbar from './Navbar';

jest.mock('react-responsive', () => {
  const originalModule = jest.requireActual('react-responsive');
  return {
    __esModule: true,
    ...originalModule,
    useMediaQuery: jest.fn(),
  };
});

describe('Navbar', () => {
  beforeEach(() => {
    useMediaQuery.mockClear();
  });

  afterAll(() => {
    useMediaQuery.mockRestore();
  });

  it('should render logo, navlinks and cart button when vw is large or more', () => {
    useMediaQuery.mockReturnValueOnce(false).mockReturnValueOnce(true);
    const renderer = new ShallowRender();
    renderer.render(<Navbar />);

    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('should render menu button, logo and cart button when vw is medium or less', () => {
    useMediaQuery.mockReturnValueOnce(true).mockReturnValueOnce(false);
    const renderer = new ShallowRender();
    renderer.render(<Navbar />);

    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('should open / close navigation menu when the user clicks the menu btn', () => {
    useMediaQuery.mockReturnValueOnce(true).mockReturnValueOnce(false);

    let menuBtn;
    let navMenuContainer;

    const wrapper = mount(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>,
    );

    useMediaQuery.mockReturnValueOnce(true).mockReturnValueOnce(false);
    menuBtn = wrapper.find('[data-testid="menu-btn"]');
    menuBtn.simulate('click');
    navMenuContainer = wrapper.find('[data-testid="navlinks-container"]');
    expect(navMenuContainer.hasClass('display')).toBe(true);

    useMediaQuery.mockReturnValueOnce(true).mockReturnValueOnce(false);
    menuBtn = wrapper.find('[data-testid="menu-btn"]');
    menuBtn.simulate('click');
    navMenuContainer = wrapper.find('[data-testid="navlinks-container"]');

    expect(navMenuContainer.hasClass('hide')).toBe(true);
  });
});
