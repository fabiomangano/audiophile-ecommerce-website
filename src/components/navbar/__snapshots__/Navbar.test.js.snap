// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render logo, navlinks and cart button when vw is large or more 1`] = `
<div
  className="container"
>
  <nav>
    <Link
      to="/"
    >
      <ForwardRef(SvgLogo) />
    </Link>
    <NavLinks
      direction="horizontal"
    />
    <span
      className="center"
    >
      <IconButton
        data-testid="cart-btn"
        onClick={[Function]}
      >
        <Icon
          color="#fff"
          direction={0}
          path="M8.625 15.833c1.132 0 2.054.935 2.054 2.084 0 1.148-.922 2.083-2.054 2.083-1.132 0-2.054-.935-2.054-2.083 0-1.15.922-2.084 2.054-2.084zm9.857 0c1.132 0 2.054.935 2.054 2.084 0 1.148-.922 2.083-2.054 2.083-1.132 0-2.053-.935-2.053-2.083 0-1.15.92-2.084 2.053-2.084zm-9.857 1.39a.69.69 0 00-.685.694.69.69 0 00.685.694.69.69 0 00.685-.694.69.69 0 00-.685-.695zm9.857 0a.69.69 0 00-.684.694.69.69 0 00.684.694.69.69 0 00.685-.694.69.69 0 00-.685-.695zM4.717 0c.316 0 .59.215.658.517l.481 2.122h16.47a.68.68 0 01.538.262c.127.166.168.38.11.579l-2.695 9.236a.672.672 0 01-.648.478H7.41a.667.667 0 00-.673.66c0 .364.303.66.674.66h12.219c.372 0 .674.295.674.66 0 .364-.302.66-.674.66H7.412c-1.115 0-2.021-.889-2.021-1.98 0-.812.502-1.511 1.218-1.816L4.176 1.32H.674A.667.667 0 010 .66C0 .296.302 0 .674 0zm16.716 3.958H6.156l1.797 7.917h11.17l2.31-7.917z"
          size={24}
        />
      </IconButton>
    </span>
  </nav>
</div>
`;

exports[`Navbar should render menu button, logo and cart button when vw is medium or less 1`] = `
<div
  className="container"
>
  <nav>
    <IconButton
      data-testid="menu-btn"
      onClick={[Function]}
    >
      <Icon
        color="#fff"
        direction={0}
        path="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        size={30}
      />
    </IconButton>
    <Link
      to="/"
    >
      <ForwardRef(SvgLogo) />
    </Link>
    <div
      className="hide"
      data-testid="navlinks-container"
    >
      <NavLinks
        direction="vertical"
      />
    </div>
    <span>
      <IconButton
        data-testid="cart-btn"
        onClick={[Function]}
      >
        <Icon
          color="#fff"
          direction={0}
          path="M8.625 15.833c1.132 0 2.054.935 2.054 2.084 0 1.148-.922 2.083-2.054 2.083-1.132 0-2.054-.935-2.054-2.083 0-1.15.922-2.084 2.054-2.084zm9.857 0c1.132 0 2.054.935 2.054 2.084 0 1.148-.922 2.083-2.054 2.083-1.132 0-2.053-.935-2.053-2.083 0-1.15.92-2.084 2.053-2.084zm-9.857 1.39a.69.69 0 00-.685.694.69.69 0 00.685.694.69.69 0 00.685-.694.69.69 0 00-.685-.695zm9.857 0a.69.69 0 00-.684.694.69.69 0 00.684.694.69.69 0 00.685-.694.69.69 0 00-.685-.695zM4.717 0c.316 0 .59.215.658.517l.481 2.122h16.47a.68.68 0 01.538.262c.127.166.168.38.11.579l-2.695 9.236a.672.672 0 01-.648.478H7.41a.667.667 0 00-.673.66c0 .364.303.66.674.66h12.219c.372 0 .674.295.674.66 0 .364-.302.66-.674.66H7.412c-1.115 0-2.021-.889-2.021-1.98 0-.812.502-1.511 1.218-1.816L4.176 1.32H.674A.667.667 0 010 .66C0 .296.302 0 .674 0zm16.716 3.958H6.156l1.797 7.917h11.17l2.31-7.917z"
          size={24}
        />
      </IconButton>
    </span>
  </nav>
</div>
`;
